optaplanner-core/src/main/java/org/optaplanner/core/impl/partitionedsearch/DefaultPartitionedSearchPhase.java

```diff
@@ -151,34 +144,61 @@ public void solve(DefaultSolverScope<Solution_> solverScope) {
                 phaseScope.setLastCompletedStepScope(stepScope);
             }
         } finally {
-            // If a partition thread throws an Exception, it is propagated here
-            // but the other partition threads won't finish any time soon, so we need to ask them to terminate
-            childThreadPlumbingTermination.terminateChildren();
+            if (Thread.interrupted()) {
+                // 1a. If current thread is interrupted, propagate interrupt signal to children by initiating
+                // abrupt shutdown.
+                executor.shutdownNow();
+            } else {
+                // 1b. Otherwise, initiate graceful shutdown of the executor. This allows partition solvers to finish
+                // solving upon detecting the termination issued in the next step (2). Shutting down the executor
+                // service is important because the JVM cannot exit until all nondaemon threads have terminated.
+                executor.shutdown();
```

Comment:
It is noteworthy that this PR reintroduces thread pool shutdown at the beginning of the finally block, which is in contradiction to e88dd1670 where it was reverted. However this time the shutdown is justified and supported by accompanying tests and explained in comments above.